python/ql/lib/semmle/python/security/dataflow/TarSlipCustomizations.qll

```diff
@@ -0,0 +1,141 @@
+
+  /**
+   * A source of exception info, considered as a flow source.
+   */
+  class TarfileOpen extends Source {
+    TarfileOpen() {
+      this = API::moduleImport("tarfile").getMember("open").getACall() and
+      // If argument refers to a string object, then it's a hardcoded path and
+      // this tarfile is safe.
+      not this.(DataFlow::CallCfgNode).getArg(0).getALocalSource().asExpr() instanceof StrConst and
+      /* Ignore opens within the tarfile module itself */
+      not this.getLocation().getFile().getBaseName() = "tarfile.py"
+    }
+  }
+
+  /**
+   * A sanitizer based on file name. This because we extract the standard library.
+   *
+   * For efficiency we don't want to track the flow of taint
+   * around the tarfile module.
+   */
+  class ExcludeTarFilePy extends Sanitizer {
+    ExcludeTarFilePy() { this.getLocation().getFile().getBaseName() = "tarfile.py" }
+  }
+
+  /**
+   * A sink capturing method calls to `extractall`.
+   *
+   * For a call to `file.extractall` without arguments, `file` is considered a sink.
+   */
+  class ExtractAllSink extends Sink {
+    ExtractAllSink() {
+      exists(DataFlow::CallCfgNode call |
+        call =
+          API::moduleImport("tarfile")
+              .getMember("open")
+              .getReturn()
+              .getMember("extractall")
+              .getACall() and
+        not exists(call.getArg(_)) and
+        not exists(call.getArgByName(_)) and
+        this = call.(DataFlow::MethodCallNode).getObject()
+      )
+    }
+  }
+
+  /**
+   * An argument to `extract` is considered a sink.
+   */
+  class ExtractSink extends Sink {
+    ExtractSink() {
+      exists(DataFlow::CallCfgNode call |
+        call =
+          API::moduleImport("tarfile").getMember("open").getReturn().getMember("extract").getACall() and
+        this = call.getArg(0)
+      )
+    }
+  }
+
+  /** The `members` argument `extractall` is considered a sink. */
+  class ExtractMembersSink extends Sink {
+    ExtractMembersSink() {
+      exists(DataFlow::CallCfgNode call |
+        call =
+          API::moduleImport("tarfile")
+              .getMember("open")
+              .getReturn()
+              .getMember("extractall")
+              .getACall() and
+        this in [call.getArg(0), call.getArgByName("members")]
+      )
+    }
+  }
+
+  /**
+   * A sanitizer guard heuristic.
+   *
+   * For a "check-like function-name" (matching `"%path"`), `checkPath`,
+   * and a call `checkPath(info.name)`, the variable `info` is considered checked.
+   */
+  class TarFileInfoSanitizer extends SanitizerGuard {
+    ControlFlowNode tarInfo;
+
+    TarFileInfoSanitizer() {
+      exists(CallNode call, AttrNode attr |
+        this = call and
+        // We must test the name of the tar info object.
+        attr = call.getAnArg() and
+        attr.getName() = "name" and
+        attr.getObject() = tarInfo
+      |
+        // Assume that any test with "path" in it is a sanitizer
```

Comment:
I think it would match something like `utilities.pathlib.filters.relative()`.

Reply:
nope, try for yourself with

```codeql
from string s
where
  s in ["ends_with_path", "path_foo"] and
  s.matches("%path")
select s
```

Reply:
Ah, right. My example should have been `utilities.path.filters.relative()`.

Reply:
Right, I'm still not 100% happy about the comment. I have two suggestions:

1. change the comment entirely, so it reads something like `// the assumption that any test that matches `%path` is a sanitizer might be too broad`
2. make the comment more precise

After having written out (1), I'm very much in favor of that version, since I guess that is what highlights the intent behind the initial comment: that the sanitizer might be too broad.

Reply:
Right, that does seem more helpful.